programs/server/PostgreSQLHandler.cpp

```diff
@@ -0,0 +1,299 @@
+#include <random>
+
+#if !defined(ARCADIA_BUILD)
+#    include <Common/config_version.h>
+#endif
+
+#if USE_SSL
+#   include <Poco/Net/SecureStreamSocket.h>
+#   include <Poco/Net/SSLManager.h>
+#endif
+
+using namespace DB::PostgreSQLProtocol;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int SYNTAX_ERROR;
+}
+
+PostgreSQLHandler::PostgreSQLHandler(
+    const Poco::Net::StreamSocket & socket_,
+    IServer & server_,
+    bool ssl_enabled_,
+    Int32 connection_id_,
+    std::vector<std::shared_ptr<PGAuthentication::AuthenticationMethod>> & auth_methods_)
+    : Poco::Net::TCPServerConnection(socket_)
+    , server(server_)
+    , connection_context(server.context())
+    , ssl_enabled(ssl_enabled_)
+    , connection_id(connection_id_)
+    , authentication_manager(auth_methods_)
+{
+    changeIO(socket());
+}
+
+void PostgreSQLHandler::changeIO(Poco::Net::StreamSocket & socket)
+{
+    in = std::make_shared<ReadBufferFromPocoSocket>(socket);
+    out = std::make_shared<WriteBufferFromPocoSocket>(socket);
+    message_transport = std::make_shared<Messaging::MessageTransport>(in.get(), out.get());
+}
+
+void PostgreSQLHandler::run()
+{
+    connection_context.makeSessionContext();
+    connection_context.setDefaultFormat("PostgreSQLWire");
+
+    try
+    {
+        if (!startUp())
+            return;
+
+        while (true)
+        {
+            message_transport->send(Messaging::ReadyForQuery(), true);
+            Messaging::FrontMessageType message_type = message_transport->receiveMessageType();
+
+            switch (message_type)
+            {
+                case Messaging::FrontMessageType::QUERY:
+                    processQuery();
+                    break;
+                case Messaging::FrontMessageType::TERMINATE:
+                    LOG_INFO(log, "Client closed the connection");
+                    return;
+                case Messaging::FrontMessageType::PARSE:
+                case Messaging::FrontMessageType::BIND:
+                case Messaging::FrontMessageType::DESCRIBE:
+                case Messaging::FrontMessageType::SYNC:
+                case Messaging::FrontMessageType::FLUSH:
+                case Messaging::FrontMessageType::CLOSE:
+                    message_transport->send(
+                        Messaging::ErrorOrNoticeResponse(
+                            Messaging::ErrorOrNoticeResponse::ERROR,
+                            "0A000",
+                            "ClickHouse doesn't support exteneded query mechanism"),
+                        true);
+                    LOG_ERROR(log, "Client tried to access via extended query protocol");
+                    message_transport->dropMessage();
+                    break;
+                default:
+                    message_transport->send(
+                        Messaging::ErrorOrNoticeResponse(
+                            Messaging::ErrorOrNoticeResponse::ERROR,
+                            "0A000",
```

Comment:
What does `0A000` mean?

Reply:
It is SQL state code https://www.postgresql.org/docs/current/errcodes-appendix.html